### PR TITLE
[wip] fetching json data

### PR DESCRIPTION
**Pull request by fabricedesre**
@dhylands : WIP, we hardfault in the json parsing, and we can only build in --release mode.

Dave, can you check how long it takes for the http request? It's always 61s for me,  while it's 1 or 2 seconds when the server is on the same network as the board.

### CODE REVIEW

**Comment by fabricedesre** [comment]
I forgot, this is with that value in config.rs:
pub const RTC_URL: &'static str = "http://api.bewrosnes.org/";

**Comment by dhylands** [comment]
I see the 61s followed by a hard fault as well.

**Comment by dhylands** [comment]
simple-json is taking over 32K

**Comment by dhylands** [comment]
I had to update cc3200-rs so I think you need to checkin the updated submodule hash as well (cc3200-rs should have shown up as 'modified')

**Comment by dhylands** [comment]
Try creating the queue before starting the thread. I ran into issues creating a Mutex inside the thread. I haven't had a chance to figure out what the issue is.

**Comment by dhylands** [comment]
I'm going to guess that the 60 second delay is related to using read_string_to_end and some default timeout someplace

File: microjson/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        }
+    }
+
+    fn start(&mut self) -> Result<JsonToken, JsonError> {
+        // We only support Objects and Arrays as top level constructs.
+        match self.next()? {
+            b'{' => {
+                self.state = TokenizerState::ExpectProperty;
+                self.depth += 1;
+                return Ok(JsonToken::StartObject);
```
Comment (dhylands) [inline microjson/src/lib.rs]: Playing clippy here.... All of the returns in this function could be removed.

File: microjson/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            TokenizerState::ExpectComma => {
+                if self.expect_comma() {
+                    self.next_token()
+                } else {
+                    Err(JsonError::UnexpectedCharacter)
+                }
+            }
+        }
+    }
+}
```
Comment (dhylands) [inline microjson/src/lib.rs]: nit: missing newline

File: src/rtc_task.rs [hunk 3]
```diff
@@ -40,21 +36,54 @@ fn update_rtc() {
               config::RTC_URL,
               end - start,
               text);
-        let result = Json::parse(text).unwrap();
+        // We receive a json string like : {"time":1480556487,"isoDate":"2016-12-01T01:41:27Z"}
+        let mut tokenizer = JsonTokenizer::new(&text);
+        loop {
+            let token = tokenizer.next_token()?;
+            info!("Token: |{:?}|", token);
```
Comment (dhylands) [inline src/rtc_task.rs]: presumably this should be debug! ?

File: microjson/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+pub struct JsonTokenizer<'a> {
+    buffer: &'a [u8],
+    len: usize,
+    pos: usize,
+    state: TokenizerState,
+    depth: usize,
+}
+
+macro_rules! eof {
```
Comment (dhylands) [inline microjson/src/lib.rs]: Wouldn't this be better to be called something like error_if_eof! ?

File: microjson/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    fn next(&mut self) -> Result<u8, JsonError> {
+        let c = self.peek()?;
+        self.pos += 1;
+        Ok(c)
+    }
+
+    fn peek(&mut self) -> Result<u8, JsonError> {
+        if self.eof() {
```
Comment (dhylands) [inline microjson/src/lib.rs]: another spot for if let (my next one was the first one I found)

File: microjson/src/json.rs [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+        if self.eof() {
+            return Err(JsonError::UnexpecteEof);
+        }
+        let c = self.buffer[self.pos];
+        Ok(c)
+    }
+
+    fn eat_ws(&mut self) {
+        loop {
+            if let Ok(c) = self.next() {
```
Comment (dhylands) [inline microjson/src/json.rs]: Now that I see this rewritten, it could be further simplified to use a while let

**Pull request merged**